### PR TITLE
OSC HA Design Proposal Document

### PR DESCRIPTION
**Pull request by bmmukund**
OSC HA Design Proposal with different approaches and future investigations.

### CODE REVIEW

**Comment by bmmukund** [comment]
@emanoelxavier,@arvindn05,@sudhirappaji ,
Please review the HA Design Proposal and let me know if any changes are required.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# OSC High Availability
+- OSC High Availability eliminates from single points of failure which protects from unplanned downtime.
+- Performs regular checks to make sure each OSC Service Node is working properly, and if one fails, switch it out for one that is working.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: if one fails, switch it out for one that is working -> If current active node moves into failure state then turn on the service in other working node.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+- It tries to bring back the failed services into the cluster and maintains the same number of services node.
+    
+#### There are 2 types of failures
+1. Service Stops
+	OSC service may be stopped abnormaly.
+2. Node Stops
+	OSC Service Node Stopped because of some reasons.
+
+## Background
+Currently, OSC does not have the HA configuration option to handle above failure types.So,A high availability OSC will work by having more service nodes than it needs,will perform regular checks to make sure each service/node is working properly, and if one fails, switching it out for one that is working.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: General Comment: maintain same format throughout the document. Give space after ",". There are other instances of this, please fix them too.

Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: if one fails, switch it out for one that is working -> If current active node moves into failure state then it turns on the service in other working node.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+We have explored two approaches.
+1. Two nodes cluster with Active/Passive setup using Pacemaker,Corosync and Distributed Replicated Block Device(DRBD).
+2. Segregate Application and Persistence Layers.
+
+## OSC HA with Active/Passive Setup 
+Two nodes cluster with Active/Passive setup using Pacemaker,Corosync and Distributed Replicated Block Device(DRBD).
+Active Node serve the requests where as Passive Node is Standby.So,on fails of Active Node,Passive Node will get promoted to Active Node.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: General Comment: Space after ".". Fix other instances of this too in the document(if any).

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+Pacemaker and Corosync maintains Service HA where It achieves maximum availability for your cluster services by detecting and recovering from node- and resource-level failures by making use of the messaging and membership capabilities provided by Corosync.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: node- -> node-level??

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+#### DRBD Architecture Diagram
+
+
+![](./images/DRBD-Arch-Diagram.png) 
+
+**DRBD is the replacement of shared storage systems by networked mirroring.Also,called Network based RAID 1 mirroring for the data.It allows read writes on Active Node where as completely disallows to Passive Node.**
+
+## OSC HA Through Segregating Application and Persistence Layers
+Currently,OSC is tightly coupled the Application and Persistence Layers.So,creating HA for persistence layer is little hard.Seprating the Application and Persistence Layers will make the OSC loose couple.Also,creating the HA for both the layers will be easier and as per the other openstack services.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: loose couple -> loosely coupled with its persistence storage??

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+1. Service Stops
+	OSC service may be stopped abnormaly.
+2. Node Stops
+	OSC Service Node Stopped because of some reasons.
+
+## Background
+Currently, OSC does not have the HA configuration option to handle above failure types.So,A high availability OSC will work by having more service nodes than it needs,will perform regular checks to make sure each service/node is working properly, and if one fails, switching it out for one that is working.
+
+We have explored two approaches.
+1. Two nodes cluster with Active/Passive setup using Pacemaker,Corosync and Distributed Replicated Block Device(DRBD).
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: Why stick to two nodes? could be multiple nodes. Try to make the statement generic.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+2. Segregate Application and Persistence Layers.
+
+## OSC HA with Active/Passive Setup 
+Two nodes cluster with Active/Passive setup using Pacemaker,Corosync and Distributed Replicated Block Device(DRBD).
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: Why stick to two nodes? could be multiple nodes. Try to make the statement generic.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+HAProxy will do the Load Balancing as well traffic redirection to the Active Node.
+
+# Concerns Using the above Proposed Designs
+1.How to Avoid Running OSC Jobs when it is in Passive Mode?
+The Passive Mode running OSC jobs may cause the problem . But we can avoid it through below strategy.
+- There should be a Monitoring thread to check the status of the node.
+- The Monitoring thread should be in the loop and in sleep state until it gets the Node status as Active.
+- The Node status can get through “pcs status resources” command.
+If it is passive , should not start the complete OSC such as RabbitMq , WebSocket , Scheduler, Opening the database connection etc. during Server Start.
+- As soon as the OSC Node gets promoted to Active , it should start remaining services.
```
Comment (sahasamiran) [inline designs/osc_high_availability/OSC_High_Availability.md]: Format problem. space before ",".

File: designs/osc_high_availability/OSC_High_Availability.md (hunk 1 above)
Comment (bmmukund) [reply designs/osc_high_availability/OSC_High_Availability.md]: Did the required changes.

**Comment by bmmukund** [comment]
@sahasamiran,
Did the required changes as per your review comments.Please review and let me know if any changes needs to be done.

**Comment by bmmukund** [comment]
@sudhirappaji,@emanoelxavier,@arvindn05,
I updated the document by adding 3rd approach.Also,formatted the issues.Please review and let me know if any changes needs to be done.

File: designs/osc_high_availability/OSC_High_Availability.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
+# OSC High Availability
```
Comment (emanoelxavier) [inline designs/osc_high_availability/OSC_High_Availability.md]: change the file name to use lower case only for consistency with the other files

**Comment by bmmukund** [comment]
@sudhirappaji,@emanoelxavier,@arvindn05,
As per Emanoel's review comment,renamed the file.Please let me know if any changes needs to be done.

**Comment by sudhirappaji** [comment]
@bmmukund , Please add point under further investigation that Option 3 needs further investigation to narrow down detailed implementation path, please add this and from my side i will approve.
@arvindn05 , Please check and feedback. I plan to merger this by tomorrow, 3/16 if no comments.

**Comment by bmmukund** [comment]
@sudhirappaji,
I added the further investigation items as suggested.Please review and let me know if any changes needs to be done.

File: designs/osc_high_availability/high_availability.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
+# OSC High Availability
+- OSC High Availability eliminates from single points of failure which protects from unplanned downtime.
+- Performs regular checks to make sure each OSC Service Node is working properly, and if one fails, switch it out for one that is working.
+- It tries to bring back the failed service/nodes into the cluster and maintains the same number of service nodes.
+    
+#### There are 2 types of failures
+1. Service Stops
+	OSC service may be stopped abnormaly.
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: typo abnormally.

Also formatting does not carry over correctly see below. "OSC service..." need to be on own line

image

File: designs/osc_high_availability/high_availability.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+We have explored three approaches.
+1. OSC node cluster with Active/Passive setup using Pacemaker, Corosync and Distributed Replicated Block Device(DRBD).
+2. Segregate Application and Persistence Layers(The option has discarded since H2 does not provide the option of Active/Passive Cluster)
+3. Using NFS and Custom Monitor Script
+
+## OSC HA with Active/Passive Setup 
+OSC node cluster with Active/Passive setup using Pacemaker, Corosync and Distributed Replicated Block Device(DRBD).
+Active Node serves the requests where as Passive Node is Standby. So, on fails of Active Node, Passive Node will get promoted to Active Node.
+Pacemaker and Corosync maintains Service HA where It achieves maximum availability for our cluster services by detecting and recovering from node- and resource-level failures by making use of the messaging and membership capabilities provided by Corosync.
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: nitpick...remove capitalization of I in "where It achieves"

File: designs/osc_high_availability/high_availability.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+## Background
+Currently, OSC does not have the HA configuration option to handle above failure types. So, A high availability OSC will work by having more service nodes than it needs, will perform regular checks to make sure each service/node is working properly. If current active node moves into failure state then it turns on the service in other working node.
+
+We have explored three approaches.
+1. OSC node cluster with Active/Passive setup using Pacemaker, Corosync and Distributed Replicated Block Device(DRBD).
+2. Segregate Application and Persistence Layers(The option has discarded since H2 does not provide the option of Active/Passive Cluster)
+3. Using NFS and Custom Monitor Script
+
+## OSC HA with Active/Passive Setup 
+OSC node cluster with Active/Passive setup using Pacemaker, Corosync and Distributed Replicated Block Device(DRBD).
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: break into logical paragraphs. Again looks like formatting was done in the raw format, but it does not translate to the final markdown output.

File: designs/osc_high_availability/high_availability.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+**DRBD is the replacement of shared storage systems by networked mirroring. Also, called Network based RAID 1 mirroring for the data. It allows read writes on Active Node where as completely disallows to Passive Node.**
+
+## OSC HA Through Segregating Application and Persistence Layers
+Currently, OSC is tightly coupled with the Application and Persistence Layers. So, creating HA for persistence layer is little hard. Seprating the Application and Persistence Layers will make the OSC loosly coupled. Also, creating the HA for both the layers will be easier and align with other openstack services.
+
+#### HA Architecture Diagram for Application and Persistence Layers
+
+![](./images/OSC-HA-Through-Segregation.png) 
+
+**Application and Persistence Layers seprated out with seprate HA. H2 Database has H2 Plugins for Active/Standby setup options and replications.**
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: typo seprated

File: designs/osc_high_availability/high_availability.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+![](./images/OSC-HA-Using-NFS.png) 
+
+**NFS Hosts File System for Docker Volumes and will be Mounted/Un-Mounted during Active/Passive.**
+
+### Deployment
+OSC cluster which will be deployed on single controller node. Threre will be a single Active instance running with extra standby node(s) for failover. Active Node always servers the request where as standby will be completely in dump state.
+
+### Load balance/HA redundancy
+HAProxy will do the Load Balancing as well traffic redirection to the Active Node.
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: dont see the HA proxy in the diagram above

File: designs/osc_high_availability/high_availability.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    - There should be a Monitoring thread to check the status of the node.
+    - The Monitoring thread should be in the loop and in sleep state until it gets the Node status as Active.
+    - The Node status can get through “pcs status resources” command.
+    If it is passive , should not start RabbitMq, WebSocket, Scheduler, Opening the database connection etc during Server Start.
+    - As soon as the OSC Node gets promoted to Active, it should start remaining OSC dependencies services such as  RabbitMq, WebSocket, Scheduler, Opening the database connection etc.
+
+2. How pacemaker notifies the status to the registered node?
+    - Pacemaker provides the brain that processes and reacts to events regarding the cluster. These events include nodes joining or leaving the cluster; resource events caused by failures, maintenance and scheduled activities; and other administrative actions. Pacemaker will compute the ideal state of the cluster and plot a path to achieve it after any of these events. This may include moving resources, stopping nodes and even forcing them offline with remote power switches.
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: break down events into bullet list?

File: designs/osc_high_availability/high_availability.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+![](./images/OSC-HA-Through-Segregation.png) 
+
+**Application and Persistence Layers seprated out with seprate HA. H2 Database has H2 Plugins for Active/Standby setup options and replications.**
+
+###### Note: Seprating Database is not feasible in OSC since H2 does not provide the Active/Passive Cluster option.Hence,discarding this option.
+
+## OSC HA Using NFS and Custom Monitoring Script
+NFS Hosts the Docker Volume File System and will be Mounted/Un-Mounted in the case Active/Passive. Custom Monitoring Script will Mount/Un-Mount the volume and Start the OSC Docker Container pointing to NFS File System. Pacemaker will execute the Custom Monitoring Script based on the status change.
+
+#### HA Architecture Diagram Using NFS and Custom Monitoring Script
```
Comment (arvindn05) [inline designs/osc_high_availability/high_availability.md]: please add more details on where the script exists, how to configure pacemaker/corosync to execute script etc.

reference to their documentation would be nice.

**Comment by bmmukund** [comment]
@arvindn05 ,
I made the changes as per your review comments.Please let me know if anything needs to be changed.

**Pull request merged**